File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/ui/activity/ShortcutActivity.java
package de.szalkowski.activitylauncher.ui.activity;

import static org.thirdparty.Launcher.launchActivity;

import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

public class ShortcutActivity extends AppCompatActivity {

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            Intent launchIntent = Intent.parseUri(getIntent().getStringExtra("extra_intent"), 0);
            launchActivity(this, launchIntent.getComponent(), false, false);
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            finish();
        }
    }
}

File: app/src/main/java/de/szalkowski/activitylauncher/ShortcutActivity.kt
package de.szalkowski.activitylauncher

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import dagger.hilt.android.AndroidEntryPoint
import de.szalkowski.activitylauncher.services.ActivityLauncherService
import de.szalkowski.activitylauncher.services.IconCreatorService
import de.szalkowski.activitylauncher.services.IntentSigningService
import javax.inject.Inject

@AndroidEntryPoint
class ShortcutActivity : AppCompatActivity() {
    @Inject
    internal lateinit var launcherService: ActivityLauncherService

    @Inject
    internal lateinit var signingService: IntentSigningService

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        try {
            val launchIntent =
                Intent.parseUri(intent.getStringExtra(IconCreatorService.INTENT_EXTRA_INTENT), 0)
            val signature =
                intent.getStringExtra(IconCreatorService.INTENT_EXTRA_SIGNATURE).orEmpty()
            val asRoot = intent.action == IconCreatorService.INTENT_LAUNCH_ROOT_SHORTCUT

            if (asRoot && !signingService.validateIntentSignature(launchIntent, signature)) {
                return
            }

            launcherService.launchActivity(
                launchIntent.component!!,
                asRoot,
                showToast = false,
            )
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            ActivityCompat.finishAffinity(this)
        }
    }
}
